File: smart_garden/lib/services/mqtt_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:mqtt_client/mqtt_client.dart';
import 'package:mqtt_client/mqtt_server_client.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import '../models/sensor_data.dart';

class MqttService {
  MqttServerClient? _client;
  
  // StreamControllers para broadcasts
  final _connectionController = StreamController<bool>.broadcast();
  final _sensorDataController = StreamController<SensorData>.broadcast();

  // Getters para os streams
  Stream<bool> get connectionStatus => _connectionController.stream;
  Stream<SensorData> get sensorDataStream => _sensorDataController.stream;

  Future<void> connect() async {
    final broker = dotenv.env['MQTT_BROKER'] ?? 'localhost';
    final port = int.parse(dotenv.env['MQTT_PORT'] ?? '1883');
    final clientId = dotenv.env['MQTT_CLIENT_ID'] ?? 'smart_garden_app';
    final topic = dotenv.env['MQTT_TOPIC_SENSORS'] ?? 'smart_garden/sensors';

    print('🔌 MQTT: Broker: $broker');
    print('🔌 MQTT: Port: $port');
    print('🔌 MQTT: Client ID: $clientId');
    print('🔌 MQTT: Conectando ao broker $broker:$port');

    _client = MqttServerClient(broker, clientId);
    _client!.port = port;
    _client!.keepAlivePeriod = 60;
    _client!.logging(on: true);
    _client!.onConnected = _onConnected;
    _client!.onDisconnected = _onDisconnected;
    _client!.onSubscribed = _onSubscribed;

    final connMessage = MqttConnectMessage()
        .withClientIdentifier(clientId)
        .startClean()
        .withWillQos(MqttQos.atLeastOnce);
    
    _client!.connectionMessage = connMessage;

    try {
      await _client!.connect();
      
      if (_client!.connectionStatus?.state == MqttConnectionState.connected) {
        print('✅ MQTT: Conectado com sucesso!');
        _connectionController.add(true);
        
        print('📥 MQTT: Inscrevendo no tópico $topic');
        _client!.subscribe(topic, MqttQos.atLeastOnce);
        
        _client!.updates!.listen((List<MqttReceivedMessage<MqttMessage>> c) {
          final MqttPublishMessage message = c[0].payload as MqttPublishMessage;
          final payload = MqttPublishPayload.bytesToStringAsString(message.payload.message);
          
          print('📨 MQTT: Mensagem recebida: $payload');
          
          try {
            final data = SensorData.fromJson(json.decode(payload));
            _sensorDataController.add(data);
            print('✅ MQTT: Dados parseados e enviados ao stream');
          } catch (e) {
            print('❌ MQTT: Erro ao parsear dados: $e');
          }
        });
      } else {
        print('🔴 MQTT: Falha na conexão');
        _connectionController.add(false);
      }
    } catch (e) {
      print('🔴 MQTT: Erro ao conectar: $e');
      _connectionController.add(false);
      rethrow;
    }
  }

  void _onConnected() {
    print('✅ MQTT: Callback onConnected');
    _connectionController.add(true);
  }

  void _onDisconnected() {
    print('⚠️ MQTT: Desconectado');
    _connectionController.add(false);
  }

  void _onSubscribed(String topic) {
    print('✅ MQTT: Inscrito no tópico: $topic');
  }

  void disconnect() {
    print('🔌 MQTT: Desconectando...');
    _client?.disconnect();
    _connectionController.close();
    _sensorDataController.close();
  }

  void publish(String topic, String message) {
    if (_client?.connectionStatus?.state == MqttConnectionState.connected) {
      final builder = MqttClientPayloadBuilder();
      builder.addString(message);
      _client!.publishMessage(topic, MqttQos.atLeastOnce, builder.payload!);
      print('📤 MQTT: Mensagem publicada em $topic: $message');
    } else {
      print('❌ MQTT: Não conectado, não foi possível publicar');
    }
  }
}

File: smart_garden/smart-garden-backend/src/services/mqttService.ts
import mqtt from 'mqtt';
import dotenv from 'dotenv';
import db from '../config/database';

dotenv.config();

class MqttService {
  private client: mqtt.MqttClient | null = null;
  private publishInterval: NodeJS.Timeout | null = null;

  async connect() {
    const broker = process.env.MQTT_BROKER || 'localhost';
    const port = process.env.MQTT_PORT || '1883';
    const brokerUrl = `mqtt://${broker}:${port}`;

    console.log('📡 MQTT Service ativo');
    console.log(`🔌 MQTT Backend: Conectando ao broker ${brokerUrl}`);

    this.client = mqtt.connect(brokerUrl);

    this.client.on('connect', () => {
      console.log('✅ MQTT Backend: Conectado ao broker com sucesso!');
      
      // Subscrever ao tópico de requests
      this.client?.subscribe('smart_garden/request', (err) => {
        if (!err) {
          console.log('📥 MQTT Backend: Inscrito em smart_garden/request');
        }
      });

      // Publicar dados a cada 5 segundos
      this.startPeriodicPublish();
    });

    this.client.on('error', (error) => {
      console.error('🔴 MQTT Backend: Erro de conexão:', error.message);
    });

    this.client.on('message', async (topic, message) => {
      console.log(`📨 MQTT Backend: Mensagem recebida em ${topic}: ${message.toString()}`);
      
      if (topic === 'smart_garden/request' && message.toString() === 'GET_CURRENT') {
        await this.publishCurrentData();
      }
    });
  }

  private startPeriodicPublish() {
    console.log('⏰ MQTT Backend: Publicação periódica iniciada (5s)');
    
    this.publishInterval = setInterval(async () => {
      await this.publishCurrentData();
    }, 5000);
  }

  private async publishCurrentData() {
    try {
      console.log('🔍 MQTT Backend: Buscando dados mais recentes da BD...');
      
      const result = await db.query(
        'SELECT * FROM dados_sensores ORDER BY data_registo DESC LIMIT 1'
      );

      if (result.rows.length > 0) {
        const data = result.rows[0];
        const payload = JSON.stringify({
          temperatura_ar: data.temperatura_ar,
          humidade_ar: data.humidade_ar,
          nivel_agua: data.nivel_agua,
          umidade_solo: data.umidade_solo,
          nivel_luz: data.nivel_luz,
          data_registo: data.data_registo
        });

        this.client?.publish('smart_garden/sensors', payload);
        console.log('📤 MQTT Backend: Dados publicados com sucesso');
      }
    } catch (error) {
      console.error('🔴 MQTT Backend: Erro ao buscar/publicar dados:', error);
    }
  }

  disconnect() {
    if (this.publishInterval) {
      clearInterval(this.publishInterval);
    }
    this.client?.end();
    console.log('🔌 MQTT Backend: Desconectado');
  }
}

export default new MqttService();